src/data/abilities/ab-attrs/bad-dreams-ab-attr.ts

```diff
@@ -0,0 +1,52 @@
+import { AbAttr } from "#abilities/ab-attr";
+import { globalScene } from "#app/global-scene";
+import { getPokemonNameWithAffix } from "#app/messages";
+import type { DrowsyTag } from "#battler-tags/drowsy-tag";
+import { AbAttrFlag } from "#enums/ab-attr-flag";
+import { BattlerTagType } from "#enums/battler-tag-type";
+import { HitResult } from "#enums/hit-result";
+import { StatusEffect } from "#enums/status-effect";
+import type { Pokemon } from "#field/pokemon";
+import { toDmgValue } from "#utils/common-utils";
+import i18next from "i18next";
+
+/**
+ * Attribute to damage all sleeping opponents by 1/8 of their max hp at the end of turn.
+ * Used for {@link https://bulbapedia.bulbagarden.net/wiki/Bad_Dreams_(Ability) | Bad Dreams}.
+ * @extends AbAttr
+ */
+export class BadDreamsAbAttr extends AbAttr {
+  constructor(showAbility: boolean = true, showAbilityInstant: boolean = false) {
+    super(showAbility, showAbilityInstant);
+    this._flags.add(AbAttrFlag.BAD_DREAMS);
+  }
+
+  public override apply(pokemon: Pokemon, simulated: boolean): boolean {
+    let isApplied = false;
+
+    for (const opp of pokemon.getOpponents()) {
+      const isAsleep = opp.hasStatusEffect(StatusEffect.SLEEP);
+      const blocksNonDirectDamage = opp.hasAbilityWithAttr(AbAttrFlag.BLOCK_NON_DIRECT_DAMAGE);
+      const willFallAsleep = opp.getTag<DrowsyTag>(BattlerTagType.DROWSY)?.turnCount === 1; // TODO: Workaround because Drowsy sets the sleep status AFTER applying bad dreams due to "asPhase = true"
```

Comment:
```suggestion
      // TODO: Workaround because Drowsy sets the sleep status AFTER applying bad dreams due to "asPhase = true"
      const willFallAsleep = opp.getTag<DrowsyTag>(BattlerTagType.DROWSY)?.turnCount === 1;
``` 

Not sure I understand what actually is to do. Do this in another way by moving the time bad dream gets applied? (Or the time drowsy get applied?)

Reply:
The time drowsy transitions over into sleep happens AFTER bad-dreams was applied

Reply:
I understand that, but not what exactly is "to do" (since it doesn't say)

Reply:
Drowsy should be setting the Sleep status BEFORE bad dreams is applied

Reply:
FYI, the additional `applyAbAttr()` call in turn-end phase with "Pokemon waking up in the same turn" has nothing to do with this one